File: expression-evaluator-test/InfixToPostfixConverterTests.cpp
#include "stdafx.h"
#include "CppUnitTest.h"
#include "../expression-evaluator/InfixToPostfixConverter.h"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace Test {
	TEST_CLASS(InfixToPostfixConverterTests) {
	public:
		TEST_METHOD(EmptyTokenList) {
			TokenList tokens;

			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::IsTrue(result.empty());
		}

		TEST_METHOD(SingleTokenInList1) {
			TokenList tokens;
			tokens.push_back(Token(3.14f));

			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::AreEqual(1, (int)result.size());
			Assert::AreEqual(3.14f, result.front().Value());
		}

		TEST_METHOD(SingleTokenInList2) {
			TokenList tokens;
			tokens.push_back(Token("test_var"));

			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::AreEqual(1, (int)result.size());
			Assert::AreEqual("test_var", result.front().Symbol().c_str());
		}

		TEST_METHOD(SimpleAddition) {
			TokenList tokens;
			tokens.push_back(Token("A"));
			tokens.push_back(Token(OperatorType::add));
			tokens.push_back(Token("B"));

			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::AreEqual(3, (int)result.size());

			auto it = result.begin();
			Assert::AreEqual("A", (*it++).Symbol().c_str());
			Assert::AreEqual("B", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::add, (int)(*it).Operator());
		}

		TEST_METHOD(MultipleOperators) {
			TokenList tokens;
			tokens.push_back(Token("A"));
			tokens.push_back(Token(OperatorType::add));
			tokens.push_back(Token("B"));
			tokens.push_back(Token(OperatorType::sub));
			tokens.push_back(Token("C"));
			
			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::AreEqual(5, (int)result.size());

			auto it = result.begin();
			Assert::AreEqual("A", (*it++).Symbol().c_str());
			Assert::AreEqual("B", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::add, (int)(*it++).Operator());
			Assert::AreEqual("C", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::sub, (int)(*it).Operator());
		}

		TEST_METHOD(MultipleOperatorsDifferentAssociativity) {
			TokenList tokens;
			tokens.push_back(Token("A"));
			tokens.push_back(Token(OperatorType::add));
			tokens.push_back(Token("B"));
			tokens.push_back(Token(OperatorType::mul));
			tokens.push_back(Token("C"));

			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::AreEqual(5, (int)result.size());

			auto it = result.begin();
			Assert::AreEqual("A", (*it++).Symbol().c_str());
			Assert::AreEqual("B", (*it++).Symbol().c_str());
			Assert::AreEqual("C", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::mul, (int)(*it++).Operator());
			Assert::AreEqual((int)OperatorType::add, (int)(*it).Operator());
		}

		TEST_METHOD(SimpleParenthesis) {
			TokenList tokens;
			tokens.push_back(Token(TokenType::LParen));
			tokens.push_back(Token("A"));
			tokens.push_back(Token(OperatorType::add));
			tokens.push_back(Token("B"));
			tokens.push_back(Token(TokenType::RParen));
			
			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::AreEqual(3, (int)result.size());

			auto it = result.begin();
			Assert::AreEqual("A", (*it++).Symbol().c_str());
			Assert::AreEqual("B", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::add, (int)(*it).Operator());
		}

		TEST_METHOD(LessSimpleParenthesis) {
			TokenList tokens;
			tokens.push_back(Token(TokenType::LParen));
			tokens.push_back(Token("A"));
			tokens.push_back(Token(OperatorType::add));
			tokens.push_back(Token("B"));
			tokens.push_back(Token(TokenType::RParen));
			tokens.push_back(Token(OperatorType::mul));
			tokens.push_back(Token(3.0f));
			
			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::AreEqual(5, (int)result.size());

			auto it = result.begin();
			Assert::AreEqual("A", (*it++).Symbol().c_str());
			Assert::AreEqual("B", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::add, (int)(*it++).Operator());
			Assert::AreEqual(3.0f, (*it++).Value());
			Assert::AreEqual((int)OperatorType::mul, (int)(*it).Operator());
		}

		TEST_METHOD(TwoSetsOfParenthesis) {
			TokenList tokens;
			tokens.push_back(Token(TokenType::LParen));
			tokens.push_back(Token("A"));
			tokens.push_back(Token(OperatorType::add));
			tokens.push_back(Token("B"));
			tokens.push_back(Token(TokenType::RParen));
			tokens.push_back(Token(OperatorType::mul));
			tokens.push_back(Token(TokenType::LParen));
			tokens.push_back(Token("A"));
			tokens.push_back(Token(OperatorType::add));
			tokens.push_back(Token("B"));
			tokens.push_back(Token(TokenType::RParen));
			
			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::AreEqual(7, (int)result.size());

			auto it = result.begin();
			Assert::AreEqual("A", (*it++).Symbol().c_str());
			Assert::AreEqual("B", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::add, (int)(*it++).Operator());
			Assert::AreEqual("A", (*it++).Symbol().c_str());
			Assert::AreEqual("B", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::add, (int)(*it++).Operator());
			Assert::AreEqual((int)OperatorType::mul, (int)(*it).Operator());
		}

		TEST_METHOD(NestedParenthesis) {
			TokenList tokens;
			tokens.push_back(Token(TokenType::LParen));
			tokens.push_back(Token("C"));
			tokens.push_back(Token(OperatorType::add));
			tokens.push_back(Token(TokenType::LParen));
			tokens.push_back(Token("A"));
			tokens.push_back(Token(OperatorType::add));
			tokens.push_back(Token("B"));
			tokens.push_back(Token(TokenType::RParen));
			tokens.push_back(Token(TokenType::RParen));
			
			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::AreEqual(5, (int)result.size());

			auto it = result.begin();
			Assert::AreEqual("C", (*it++).Symbol().c_str());
			Assert::AreEqual("A", (*it++).Symbol().c_str());
			Assert::AreEqual("B", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::add, (int)(*it++).Operator());
			Assert::AreEqual((int)OperatorType::add, (int)(*it).Operator());
		}

		TEST_METHOD(DifferentAssociations) {
			TokenList tokens;
			tokens.push_back(Token(OperatorType::neg));
			tokens.push_back(Token("A"));
			tokens.push_back(Token(OperatorType::add));
			tokens.push_back(Token("B"));

			std::map<OperatorType, std::pair<Associativity, Precedence>> opMap;
			opMap[OperatorType::add] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::sub] = std::make_pair(Associativity::left, Precedence::low);
			opMap[OperatorType::neg] = std::make_pair(Associativity::right, Precedence::mid);
			opMap[OperatorType::mul] = std::make_pair(Associativity::left, Precedence::high);
			opMap[OperatorType::div] = std::make_pair(Associativity::left, Precedence::high);

			InfixToPostfixConverter converter(opMap);

			TokenList result = converter(tokens);

			Assert::AreEqual(4, (int)result.size());

			auto it = result.begin();
			Assert::AreEqual("A", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::neg, (int)(*it++).Operator());
			Assert::AreEqual("B", (*it++).Symbol().c_str());
			Assert::AreEqual((int)OperatorType::add, (int)(*it).Operator());
		}
	};
}

File: Source/fireland15.LindenmayerSystem/fireland15.LindenmayerSystem.Test.Expressions/InfixToPostfixConverterTests.cs
using fireland15.LindenmayerSystem.Expressions;
using fireland15.LindenmayerSystem.Expressions.Model.Tokens;
using System.Collections.Generic;
using Xunit;

namespace fireland15.LindenmayerSystem.Test.Expressions
{
    public class InfixToPostfixConverterTests
    {
        [Fact]
        public void EmptyTokenList()
        {
            // arrange
            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(new List<Token>());

            // assert
            Assert.NotNull(result);
            Assert.Empty(result);
        }

        [Fact]
        public void SingleTokenInList1()
        {
            // arrange
            var tokens = new List<Token>
            {
                new Token(3.14159f)
            };
            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(tokens);

            // assert
            Assert.Single(result);
            Assert.Equal(tokens[0], result[0]);
        }

        [Fact]
        public void SingleTokenInList2()
        {
            // arrange
            var tokens = new List<Token>
            {
                new Token("bananas")
            };
            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(tokens);

            // assert
            Assert.Single(result);
            Assert.Equal(tokens[0], result[0]);
        }

        [Fact]
        public void SingleOperator()
        {
            // arrange
            var tokens = new List<Token>
            {
                new Token("bananas"),
                new Token(Operator.Add),
                new Token("apples")
            };
            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(tokens);

            // assert
            Assert.Equal(3, result.Count);
            Assert.Equal(tokens[0], result[0]);
            Assert.Equal(tokens[2], result[1]);
            Assert.Equal(tokens[1], result[2]);
        }

        [Fact]
        public void MultipleOperators()
        {
            // arrange
            var tokens = new List<Token>
            {
                new Token("bananas"),
                new Token(Operator.Add),
                new Token("apples"),
                new Token(Operator.Subtract),
                new Token("carrots")
            };
            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(tokens);

            // assert
            Assert.Equal(5, result.Count);
            Assert.Equal(tokens[0], result[0]);
            Assert.Equal(tokens[2], result[1]);
            Assert.Equal(tokens[1], result[2]);
            Assert.Equal(tokens[4], result[3]);
            Assert.Equal(tokens[3], result[4]);
        }

        [Fact]
        public void MultipleOperatorsDifferentPrecedence()
        {
            // arrange
            var tokens = new List<Token>
            {
                new Token("bananas"),
                new Token(Operator.Add),
                new Token("apples"),
                new Token(Operator.Multiply),
                new Token("carrots")
            };
            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(tokens);

            // assert
            Assert.Equal(5, result.Count);
            Assert.Equal(tokens[0], result[0]);
            Assert.Equal(tokens[2], result[1]);
            Assert.Equal(tokens[4], result[2]);
            Assert.Equal(tokens[3], result[3]);
            Assert.Equal(tokens[1], result[4]);
        }

        [Fact]
        public void SimpleParenthesis()
        {
            // arrange
            var tokens = new List<Token>
            {
                new Token(TokenType.LeftParenthesis),
                new Token("bananas"),
                new Token(Operator.Add),
                new Token("apples"),
                new Token(TokenType.RightParenthesis)
            };
            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(tokens);

            // assert
            Assert.Equal(3, result.Count);
            Assert.Equal(tokens[1], result[0]);
            Assert.Equal(tokens[3], result[1]);
            Assert.Equal(tokens[2], result[2]);
        }

        [Fact]
        public void LessSimpleParenthesis()
        {
            // arrange
            var tokens = new List<Token>
            {
                new Token(TokenType.LeftParenthesis),
                new Token("bananas"),
                new Token(Operator.Add),
                new Token("apples"),
                new Token(TokenType.RightParenthesis),
                new Token(Operator.Multiply),
                new Token(3.0f)
            };
            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(tokens);

            // assert
            Assert.Equal(5, result.Count);
            Assert.Equal(tokens[1], result[0]);
            Assert.Equal(tokens[3], result[1]);
            Assert.Equal(tokens[2], result[2]);
            Assert.Equal(tokens[6], result[3]);
            Assert.Equal(tokens[5], result[4]);
        }

        [Fact]
        public void TwoSetsOfParenthesis()
        {
            // arrange
            var tokens = new List<Token>
            {
                new Token(TokenType.LeftParenthesis),
                new Token("bananas"),
                new Token(Operator.Add),
                new Token("apples"),
                new Token(TokenType.RightParenthesis),
                new Token(Operator.Multiply),
                new Token(TokenType.LeftParenthesis),
                new Token("carrots"),
                new Token(Operator.Add),
                new Token("dogs"),
                new Token(TokenType.RightParenthesis),
            };
            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(tokens);

            // assert
            Assert.Equal(7, result.Count);
            Assert.Equal(tokens[1], result[0]);
            Assert.Equal(tokens[3], result[1]);
            Assert.Equal(tokens[2], result[2]);
            Assert.Equal(tokens[7], result[3]);
            Assert.Equal(tokens[9], result[4]);
            Assert.Equal(tokens[8], result[5]);
            Assert.Equal(tokens[5], result[6]);
        }

        [Fact]
        public void NestedParenthesis()
        {
            // arrange
            var tokens = new List<Token>
            {
                new Token(TokenType.LeftParenthesis),
                new Token("apples"),
                new Token(Operator.Add),
                new Token(TokenType.LeftParenthesis),
                new Token("bananas"),
                new Token(Operator.Add),
                new Token("carrots"),
                new Token(TokenType.RightParenthesis),
                new Token(TokenType.RightParenthesis),
            };

            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(tokens);

            // assert
            Assert.Equal(5, result.Count);
            Assert.Equal(tokens[1], result[0]);
            Assert.Equal(tokens[4], result[1]);
            Assert.Equal(tokens[6], result[2]);
            Assert.Equal(tokens[5], result[3]);
            Assert.Equal(tokens[2], result[4]);
        }

        [Fact]
        public void DifferentAssociations()
        {
            // arrange
            var tokens = new List<Token>
            {
                new Token(Operator.Negate),
                new Token("apples"),
                new Token(Operator.Add),
                new Token("bananas"),
            };

            var converter = new InfixToPostfixConverter();

            // act
            var result = converter.Convert(tokens);

            // assert
            Assert.Equal(4, result.Count);
            Assert.Equal(tokens[1], result[0]);
            Assert.Equal(tokens[0], result[1]);
            Assert.Equal(tokens[3], result[2]);
            Assert.Equal(tokens[2], result[3]);
        }
    }
}
